### DISCUSSION TITLE
How values are extracted from returned rows

### DISCUSSION THREAD

**Discussion by AbdulWMughal** [c410-f3r/wtx#192] [General]
Hi,

I’d like to know how WTX extracts values from returned rows. Does it use a linear scan?

Thanks

**Comment by c410-f3r** [comment]
The TCP connection receives data from the database and puts it into a buffer. This buffer can contain, among other things, the rows of a query.

```
             Buffer
             _______
TCP data -> | Row 1 |
TCP data -> | Row 2 |
TCP data -> | Row 3 |
TCP data -> | Row 4 |
            |_______|
```

High-level functions like `fetch_many_with_stmt` simply tells you where each row is located within the buffer.

> Does it use a linear scan?

As stated above, `wtx` just copies what the server sent so there isn't a "linear scan" behind the scenes. However, a linear iteration of rows can still happen if you explicitly call `iter()`.

```rust
let records = executor
  .fetch_many_with_stmt("SELECT id FROM some_table", (), |_| Ok(()))
  .await?;

// Linear scan of rows
for record in records.iter() {
  dbg!(record.decode::<_, u32>("id"));
}
```

Take a look at https://www.postgresql.org/docs/16/protocol.html to see how PostgreSQL sends bytes to clients.

**Comment by AbdulWMughal** [comment]
This benchmark https://github.com/diesel-rs/metrics/ is for the performance of extracting values from the returned rows by numeric indices or column name for each of the crates. For WTX it doesn't specify.

For extracting values by column’s name’s rather than numeric indices Tokio_Postgres uses a linear scan by default for this. As you can see extracting by column name for Tokio_Posgres has really poor performance. A hash map from name to index might improve performance but even that would depend on the number of columns in the rows. To fundamentally maximise performance with this library, you are fundamentally going to have to use numeric indices.

Do you have any insight based on your own experience and implementations with WTX on what small or major changes may be needed to remedy this issue.

Kind Regards

**Comment by c410-f3r** [comment]
Looks like `tokio_postgres` is indeed doing a linear search when dealing with strings -> https://github.com/sfackler/rust-postgres/blob/257bcfd283479de024330bc6e8919d61da5e2b98/tokio-postgres/src/row.rs#L63-L80

You probably shouldn't be too concern about micro benchmarks. For example, the "Medium Complex Query" benchmark is measuring implementations in nanoseconds so score differences likely won't matter much in production workloads.

> Do you have any insight based on your own experience and implementations with WTX on what small or major changes may be needed to remedy this issue.

You could perhaps create a PR in the `tokio_postgres` repository introducing the usage of `HashMap` in the codebase. Not sure if it is possible to speed up string look ups in your code alone.
